File: src/dataset/javascript/simple-linked-list_exercism-javascript-ffflorian/claude/ClaudeSonnet4_simple-linked-list_v2.js
export class Element {
  constructor(value = null, next = null) {
    this.value = value;
    this.next = next;
  }
}

export class List {
  constructor(values = []) {
    this.head = null;
    this._length = 0;
    for (let i = values.length - 1; i >= 0; i--) {
      this.add(new Element(values[i]));
    }
  }

  *[Symbol.iterator]() {
    let pointer = this.head;
    while (pointer) {
      yield pointer;
      pointer = pointer.next;
    }
  }

  add(element) {
    if (this.head === null) {
      this.head = element;
    } else {
      element.next = this.head;
      this.head = element;
    }
    this._length++;
  }

  reverse() {
    const values = [];
    let pointer = this.head;
    while (pointer) {
      values.push(pointer.value);
      pointer = pointer.next;
    }
    return new List(values);
  }

  toArray() {
    const result = [];
    let pointer = this.head;
    while (pointer) {
      result.push(pointer.value);
      pointer = pointer.next;
    }
    return result;
  }

  get length() {
    return this._length;
  }
}

File: src/dataset/go/simple_linked_list_Exercism_go_drapala/claude/ClaudeSonnet4_simple_linked_list_v2.go
package linkedlist

import "errors"

type Element struct {
	data int
	next *Element
}

type List struct {
	head *Element
	tail *Element
	size int
}

func New(elements []int) *List {
	if len(elements) == 0 {
		return &List{}
	}
	
	head := &Element{data: elements[0]}
	current := head
	
	for i := 1; i < len(elements); i++ {
		current.next = &Element{data: elements[i]}
		current = current.next
	}
	
	return &List{
		head: head,
		tail: current,
		size: len(elements),
	}
}

func (l *List) Size() int {
	return l.size
}

func (l *List) Push(element int) {
	newElement := &Element{data: element}
	
	if l.head == nil {
		l.head = newElement
		l.tail = newElement
	} else {
		l.tail.next = newElement
		l.tail = newElement
	}
	
	l.size++
}

func (l *List) Pop() (int, error) {
	if l.head == nil {
		return 0, errors.New("list is empty")
	}
	
	if l.head == l.tail {
		data := l.head.data
		l.head = nil
		l.tail = nil
		l.size = 0
		return data, nil
	}
	
	current := l.head
	for current.next != l.tail {
		current = current.next
	}
	
	data := l.tail.data
	l.tail = current
	l.tail.next = nil
	l.size--
	
	return data, nil
}

func (l *List) Array() []int {
	if l.size == 0 {
		return []int{}
	}
	
	result := make([]int, 0, l.size)
	current := l.head
	
	for current != nil {
		result = append(result, current.data)
		current = current.next
	}
	
	return result
}

func (l *List) Reverse() *List {
	if l.head == nil || l.head.next == nil {
		return &List{head: l.head, tail: l.tail, size: l.size}
	}
	
	var prev *Element
	current := l.head
	newTail := l.head
	
	for current != nil {
		next := current.next
		current.next = prev
		prev = current
		current = next
	}
	
	return &List{
		head: prev,
		tail: newTail,
		size: l.size,
	}
}